### PR TITLE
Extended Syntax: case hoisting

### PR DESCRIPTION
**Pull request by Anabra**
See #32

### CODE REVIEW

File: grin/src/Transformations/ExtendedSyntax/Optimising/CaseHoisting.hs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+import Data.Bifunctor (first)
+
+import Grin.ExtendedSyntax.Grin
+import Grin.ExtendedSyntax.TypeEnv
+import Transformations.ExtendedSyntax.Util
+import Transformations.ExtendedSyntax.Names
+
+{-
+  IDEA:
+    If Alt had name then the HPT could calculate it's return type and store in TypeEnv
```
Comment (andorp) [inline grin/src/Transformations/ExtendedSyntax/Optimising/CaseHoisting.hs:24]: Alt has a name now. Maybe create an issue about it?

File: grin/src/Transformations/ExtendedSyntax/Optimising/CaseHoisting.hs [hunk 2]
```diff
@@ -80,9 +80,8 @@ hoistAlts :: Name -> (Alt, Alt) -> NameM Alt
 hoistAlts lpatName (Alt cpat1 altName1 alt1, Alt cpat2 altName2 alt2) = do
   freshLPatName <- deriveNewName lpatName
   let nameMap = Map.singleton lpatName freshLPatName
-  (freshAlt2, _) <- case cpat2 of
-    DefaultPat  -> refreshNames nameMap alt2
-    _           -> refreshNames nameMap $ EBind (SReturn $ Var freshLPatName) (VarPat altName2) alt2
+  (freshAlt2, _) <- refreshNames nameMap $
```
Comment (andorp) [inline grin/src/Transformations/ExtendedSyntax/Optimising/CaseHoisting.hs:83]: What was the problem here?

Comment (Anabra) [reply grin/src/Transformations/ExtendedSyntax/Optimising/CaseHoisting.hs:83]: Now the default alt also has a name. It has to be rebound when hoisting the default alt into some other alt. See the `default pattern` test case.

File: grin/src/Transformations/ExtendedSyntax/Optimising/CaseHoisting.hs (hunk 1 above)
Comment (Anabra) [reply grin/src/Transformations/ExtendedSyntax/Optimising/CaseHoisting.hs:24]: Well, currently HPT tracks the alt name as a restricted version of the scrutinee. I will create an issue just for tracking this.

Comment (Anabra) [reply grin/src/Transformations/ExtendedSyntax/Optimising/CaseHoisting.hs:24]: See #94

**Pull request merged**